Framework/Core/System/IO/StreamReader.cs

```diff
@@ -87,7 +87,7 @@ public override int Peek()
             int nextChar;
 
             // If buffer need refresh take into account max UTF8 bytes if the next character is UTF8 encoded
-            if ((m_curBufPos                       == (m_curBufLen-1)) || 
+            if ((m_curBufPos                       >= (m_curBufLen-1)) || 
                ((m_buffer[m_curBufPos + 1] & 0x80) !=  0 && 
                 (m_curBufPos + 3                   >=  m_curBufLen)))

```

Comment:
The problem is that the read method increment the position pointer on every read. On the next read, if the position is out of range, it reload the internal buffer. This works fine expect one specific scenario where \r is at the end of buffer and StreamReader wants to peek the next character (to see if the next character is \n). In peek method if we have reached the end, the Peek method tries to reload the internal buffer. To see if the end is reached, the method use m_curBufPos == (m_curBufLen-1) but it does not work if the case where the end is passed.
This change fix the issue by changing just this reloading logic in Peek method. Alternatively, I could change the logic of read method. However, this change is simpler and does not have any side effect. So I prefer this fix.